src/vehicle.cpp

```diff
@@ -1613,20 +1614,39 @@ void Vehicle::UpdateViewport(bool dirty)
 	new_coord.right  += pt.x + 2 * ZOOM_LVL_BASE;
 	new_coord.bottom += pt.y + 2 * ZOOM_LVL_BASE;
 
-	UpdateVehicleViewportHash(this, new_coord.left, new_coord.top);
+	if (update_cache) {
+		/*
+		 * If the old coordinates are invalid, set the cache to the new coordinates for correct
+		 * behaviour the next time the coordinate cache is checked.
+		 */
+		this->sprite_cache.old_coord = this->coord.left == INVALID_COORD ? new_coord : this->coord;
+	}
 
-	Rect old_coord = this->coord;
 	this->coord = new_coord;
+}
+
+/**
+ * Update the vehicle on the viewport, updating the right hash and setting the
+ *  new coordinates.
+ * @param dirty Mark the (new and old) coordinates of the vehicle as dirty.
+ */
+void Vehicle::UpdateViewport(bool dirty)
+{
+	Rect old_coord;
+	old_coord = this->sprite_cache.old_coord;
```

Comment:
No need to split this

Reply:
actually, is this even necessary anymore? Any reason not to just use this->sprite_cache.old_coord directly?

Reply:
The sprite cache will be updated by `UpdateBoundingBoxCoordinates`, and we want the value from before this update was made to be able to identify the full extent of the dirty region. I'll combine the declaration/assignment though, that looks a lot cleaner.